### PR TITLE
Remove `unsafe` modifier from `Memory<T>.Span`

### PR DESCRIPTION
**Pull request by xtqqczze**

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Memory.cs [hunk 1]
```diff
@@ -343,7 +343,7 @@ public unsafe Span<T> Span
                     }
 #endif
 
-                    refToReturn = ref Unsafe.Add(ref refToReturn, (IntPtr)(void*)desiredStartIndex);
+                    refToReturn = ref Unsafe.Add(ref refToReturn, (nint)desiredStartIndex);
```
Comment (Sergio0694) [inline src/libraries/System.Private.CoreLib/src/System/Memory.cs:346]: How come we need to cast to `nint` in the first place? Doesn't `Unsafe.Add` have a `nuint` overload now? 🤔

Comment (xtqqczze) [reply src/libraries/System.Private.CoreLib/src/System/Memory.cs:346]: Seems to be a pattern in `System.Private.CoreLib`. Possibly `Add<T>(T, UIntPtr)` is not handled in an intrinsic on some platforms?

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Memory.cs:346]: It is more likely left-over from the days when IntPtr/UIntPtr were not regular integers in C#.

**Comment by xtqqczze** [comment]
@MihuBot

File: src/libraries/System.Private.CoreLib/src/System/Memory.cs [hunk 2]
```diff
@@ -327,12 +327,14 @@ public unsafe Span<T> Span
...
 
+                    // We use 'nuint' because it gives us a free early zero-extension to 64 bits when running on a 64-bit platform.
                     nuint desiredStartIndex = (uint)_index & (uint)ReadOnlyMemory<T>.RemoveFlagsBitMask;
+
                     int desiredLength = _length;
 
 #if TARGET_64BIT
                     // See comment in Span<T>.Slice for how this works.
-                    if ((ulong)desiredStartIndex + (ulong)(uint)desiredLength > (ulong)(uint)lengthOfUnderlyingSpan)
+                    if ((ulong)(uint)desiredStartIndex + (ulong)(uint)desiredLength > (ulong)(uint)lengthOfUnderlyingSpan)
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Memory.cs:337]: The cast to `uint` is undesirable here since `desiredStartIndex` is `nuint`.

**Review by stephentoub — APPROVED** [review]

**Comment by xtqqczze** [comment]
@stephentoub I can't see how the build analysis failure is related.

**Review by GrabYourPitchforks — APPROVED** [review]

**Pull request merged**